synapse/events/snapshot.py

```diff
@@ -254,6 +256,83 @@ async def get_prev_state_ids(
+    also may be the information after the event, if the event is not a state event). The
+    UnpersistedEventContext must be converted into an EventContext by calling the method
+    'persist' on it before it is suitable to be sent to the DB for processing.
+
+        state_handler:
+            an instance of the class StateHandler, this is required for all
+            but outlier instances
+        state_group_before_event:
+            the state group at/before the event. This is required if `for_batch` is True.
+        state_map:
+             the current state of the room
+        for_batch:
+            whether the context is part of a group of events being created for batch persisting
+            to the DB.
+        app_service: If this event is being sent by a (local) application service, that
+            app service.
+
+    """
+    _storage: "StorageControllers"
+    state_handler: Optional["StateHandler"] = None
+    state_group_before_event: Optional[int] = None
+    state_map: Optional[StateMap[str]] = None
+    for_batch: bool = False
+    app_service: Optional[ApplicationService] = None
+
+    @staticmethod
+    def for_outlier(
+        storage: "StorageControllers",
+    ) -> "UnpersistedEventContext":
+        """Returns an UnpersistedEventContext instance suitable for an outlier event"""
+        return UnpersistedEventContext(storage=storage)
+
+    async def persist(
+        self,
+        event: EventBase,
+    ) -> "EventContext":
+        """
+        Converts an UnpersistedEventContext into an EventContext, which is now suitable for
+        sending to the DB along with the event.
+        """
+        if event.internal_metadata.is_outlier():
+            return EventContext.for_outlier(self._storage)
+
+        assert self.state_handler is not None
+        if self.for_batch:
+            assert self.state_group_before_event is not None
+            assert self.state_map is not None
+            return await self.state_handler.compute_event_context_for_batched(
+                event, self.state_map, self.state_group_before_event
+            )
+        elif event.type == EventTypes.MSC2716_INSERTION:
+            return await self.state_handler.compute_event_context(
+                event, state_ids_before_event=self.state_map, partial_state=False
+            )
+        return await self.state_handler.compute_event_context(event)
+
+    async def get_prev_state_ids(
+        self, state_filter: Optional["StateFilter"] = None
+    ) -> StateMap[str]:
+        """
+        Returns the state at/before the event (which also may be the same as after the event
+        if the event is not a state event.)
+        """
+        if self.state_map is not None:
+            return self.state_map
+        else:
+            assert self.state_group_before_event is not None
+            return await self._storage.state.get_state_ids_for_group(
+                self.state_group_before_event, state_filter
+            )
```

Comment:
Hmm I think I am a little confused by your question, would you mind rephrasing?

Reply:
Sure. If the `UnpersistedEventContext` already has a `state_map` attribute, we return it immediately on line +328. If not, we hit this `else` branch. It looks to me like we're computing a state map here. That involves some work: in the worst case, db queries.

If we were to call `get_prev_state_ids` again, we'd have to repeat that work (although caching would likely kick in, so it's unlikely to be expensive). We could avoid this by writing

```python
            self.state_map = await self._storage.state.get_state_ids_for_group(
                self.state_group_before_event, state_filter
            )
            return self.state_map
```

at which point there could be just one `return` statement. 

I'm not sure if that actually gains us much though, because a) we may not call this function twice, and b) if we do, caching probably helps us anyway.